text/3727-unsafe-set-discriminant.md

```diff
@@ -0,0 +1,299 @@
+let init_some: Option<u32> = get_user_input();
+
+if let Some(val) = init_some {
+    let opt_ptr: *mut Option<[u32; 1024]> = out.as_mut_ptr();
+    // Tracking issue #120141 for enum variant offset_of! definition
+    let val_offset: usize = offset_of!(Option<[u32; 1024]>, Some.0);
+    let arr_ptr: *mut [u32; 1024] = opt_ptr.byte_add(val_offset).cast();
+    let item_ptr: *mut u32 = arr_ptr.cast();
+    // Initialize all items in the array to the user provided value
+    for i in 0..1024 {
+        // SAFETY: The variant body is always well aligned and valid for
+        // the size of the type, uninit fields are only written.
+        unsafe {
+            item_ptr.add(i).write(val);
+        }
+    }
+    // Set the discriminant
+    let discrim = discriminant_of!(Option<[u32; 1024]>, Some);
+    // SAFETY: We have initialized all fields for this variant, and
+    // this discriminant is correct for the type we are writing to.
+    unsafe {
+        set_discriminant(&mut out, discrim);
+    }
+} else {
+    // No value to write, just set the discriminant, leaving the
+    // rest of the value uninitialized
+    //
+    // SAFETY: We have initialized all fields for this variant
+    // (which is 'no fields'), and this discriminant is correct
+    // for the type we are writing to.
+    unsafe {
+        // We can also use discriminant_of! without binding it
+        set_discriminant(
+            &mut out,
+            discriminant_of!(Option<[u32; 1024]>, None),
+        );
+    }
+}
+
+// This is now sound. We could also use `assume_init_ref` or
+// `assume_init_mut` if we are explicitly avoiding potential
+// copies by value, and the allocation is not local.
+let out: Option<[u32; 1024]> = unsafe { out.assume_init() };
+assert_eq!(out.is_some(), init_some.is_some());
+if let Some(val) = init_some {
+    assert!(out.as_ref().unwrap().iter().all(|x| *x == val));
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## The `discriminant_of!` macro
+
+This macro is the "valueless" version of `std::mem::discriminant`. It needs to be a macro, as we cannot otherwise name the variant.
+
+Unlike the `discriminant` function, it is possible at compile time to detect if this is called on a type that is not an `enum`, meaning that it is possible for it to be a compiler error in this case, rather than "unspecified but not UB" to use this with a non-enum type.
+
+This macro returns by value the same opaque `Discriminant<T>` used by `std::mem::discriminant`.
+
+This RFC does not propose allowing obtaining the discriminant of "nested" fields, and will only work with a "top level" item, but does not preclude adding this capability in the future. This addition could be specified in a later RFC. For now users are recommended to handle "one level at a time".
+
+```rust
+discriminant_of!(Option<Option<u32>>, Some);        // Allowed
+discriminant_of!(Option<Option<u32>>, Some.0.Some); // Not Allowed (yet)
```

Comment:
How exactly would this work, anyway? 

The presumption here is that `Discriminant<T>` only directly covers the discriminants of the type `T`, and not its children, so, I'm not even sure what this kind of expression would return. Would it return `Discriminant<Option<u32>>`?

Reply:
Yep, I don't want to get too much into this (I wanted to leave it out for simplicity), but I would expect that second one to return `Discriminant<Option<u32>>`.

Reply:
I mostly point it out since it feels disingenuous to say it's simply "not allowed yet" and rather that it doesn't make as much sense as an operation, and would need a proper design first.

Reply:
I think we're on the same page! I'm saying "the implementor of this macro doesn't need to worry about that possibility, as we haven't specified it outside of 'don't do that'".

If you can think of a better way to word that to impart the meaning of "not defined within the scope of this RFC", I'm open to change it!